src/contracts/Output/VisitorAdapterNormalizer.php

```diff
@@ -0,0 +1,178 @@
+use Symfony\Component\Serializer\Normalizer\NormalizerAwareTrait;
+use Symfony\Component\Serializer\Normalizer\NormalizerInterface;
+
+final class VisitorAdapterNormalizer implements NormalizerInterface, NormalizerAwareInterface, ContextAwareNormalizerInterface
+{
+    use NormalizerAwareTrait;
+
+    private const string CALLED_CONTEXT = __CLASS__ . '_CALLED';
+
+    public const string ENCODER_CONTEXT = 'ENCODER_CONTEXT';
+
+    public function __construct(
+        private readonly EncoderInterface $jsonEncoder,
+        private readonly EncoderInterface $xmlEncoder,
+        private readonly ValueObjectVisitorResolverInterface $valueObjectVisitorResolver,
+    ) {
+    }
+
+    /**
+     * @param array<string, mixed> $context
+     *
+     * @throws \LogicException
+     *
+     * {@inheritDoc}
+     */
+    public function normalize(mixed $object, ?string $format = null, array $context = []): mixed
+    {
+        $eligibleVisitor = is_object($object)
+            ? $this->valueObjectVisitorResolver->resolveValueObjectVisitor($object)
+            : null;
+
+        if ($eligibleVisitor instanceof ValueObjectVisitor) {
+            return $this->visitValueObject($object, $eligibleVisitor, $format, $context);
+        }
+
+        return $this->normalizer->normalize($object, $format, $context);
+    }
+
+    /**
+     * @param array<string, mixed> $context
+     */
+    public function supportsNormalization(mixed $data, ?string $format = null, array $context = []): bool
+    {
+        if (($context[self::CALLED_CONTEXT] ?? false) === true) {
+            return false;
+        }
+
+        $eligibleVisitor = is_object($data)
+            ? $this->valueObjectVisitorResolver->resolveValueObjectVisitor($data)
+            : null;
+
+        if ($eligibleVisitor instanceof ValueObjectVisitor) {
+            return true;
+        }
+
+        if (!$this->normalizer instanceof ContextAwareNormalizerInterface) {
+            throw new LogicException(
+                sprintf(
+                    'Normalizer "%s" must be an instance of "%s".',
+                    $this->normalizer::class,
+                    ContextAwareNormalizerInterface::class,
+                ),
+            );
+        }
+
+        return $this->normalizer->supportsNormalization(
+            $data,
+            $format,
+            $context + [self::CALLED_CONTEXT => true],
+        );
+    }
+
+    /**
+     * @param array<mixed> $context
+     *
+     * @return array<mixed>
+     *
+     * @throws \LogicException
+     */
+    private function visitValueObject(
+        object $object,
+        ValueObjectVisitor $valueObjectVisitor,
+        ?string $format,
+        array $context,
+    ): array {
+        $visitor = $context['visitor'] ?? $this->createVisitor($format);
+        $generator = $visitor->getGenerator();
+
+        $generator->reset();
+        $generator->startDocument($object);
+
+        $valueObjectVisitor->visit($visitor, $generator, $object);
+
```

Comment:
```suggestion
```

Reply:
I would rather have the newline here as we have new line after `startDocument`, it's clearer to read.